File: whoathere/helpers/macos-vm-helper/Sources/WhoaThereMacosVmHelperCore/LinuxVzNetworkEvidencePayload.swift
import Foundation

public let linuxVzNetworkEvidencePayloadSchemaV1 =
    "whoathere.linux_vz_network_evidence_payload.v1"
public let linuxVzNetworkEvidenceSerialPrefixV1 = Data(
    "WHOATHERE_GUEST_NETWORK_EVIDENCE ".utf8
)
public let maximumLinuxVzNetworkEvidencePayloadBytesV1 = 64 * 1024

public enum LinuxVzNetworkEvidencePayloadError: Error, Equatable {
    case missing
    case duplicate
    case limitExceeded
    case invalidJSON
    case nonCanonical
    case invalidSchema
    case invalidEvent
}

public struct LinuxVzNetworkEvidencePayload: Equatable, Sendable {
    public let canonicalJSON: Data
    public let payloadSHA256: String
    public let evidenceByteLength: UInt64
    public let fixtureCase: String
    public let packageUID: UInt64
    public let packageGID: UInt64
    public let sourcePort: UInt16
    public let claims: LinuxVzTelemetryGuestObservationClaims
}

public func decodeLinuxVzNetworkEvidencePayloadV1(
    _ serial: Data
) throws -> LinuxVzNetworkEvidencePayload {
    var payload: Data?
    for rawLine in serial.split(separator: 0x0a, omittingEmptySubsequences: false) {
        var line = Data(rawLine)
        if line.last == 0x0d { line.removeLast() }
        guard line.starts(with: linuxVzNetworkEvidenceSerialPrefixV1) else { continue }
        guard payload == nil else { throw LinuxVzNetworkEvidencePayloadError.duplicate }
        payload = line.dropFirst(linuxVzNetworkEvidenceSerialPrefixV1.count)
    }
    guard let payload else { throw LinuxVzNetworkEvidencePayloadError.missing }
    return try decodeLinuxVzNetworkEvidenceJSONV1(payload)
}

public func decodeLinuxVzNetworkEvidenceJSONV1(
    _ data: Data
) throws -> LinuxVzNetworkEvidencePayload {
    guard !data.isEmpty, data.count <= maximumLinuxVzNetworkEvidencePayloadBytesV1 else {
        throw LinuxVzNetworkEvidencePayloadError.limitExceeded
    }
    guard let value = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
        throw LinuxVzNetworkEvidencePayloadError.invalidJSON
    }
    guard try canonicalJSONData(value) == data else {
        throw LinuxVzNetworkEvidencePayloadError.nonCanonical
    }
    let expectedKeys = Set([
        "descendant_teardown_complete", "dropped_event_count", "event_count",
        "event_sequence_end", "event_sequence_start", "events", "evidence_truncated",
        "fixture_case", "heartbeat_count", "network_action", "network_family",
        "network_protocol", "network_socket_state", "network_source", "network_source_port",
        "network_target", "network_target_port", "package_gid", "package_uid",
        "reaped_process_count", "schema_version", "sensor_healthy"
    ])
    let fixtureCase: String
    switch (
        value["fixture_case"] as? String,
        value["network_family"] as? String,
        value["network_source"] as? String,
        value["network_target"] as? String
    ) {
    case ("ipv4_connect", "ipv4", "192.0.2.2", "192.0.2.1"):
        fixtureCase = "ipv4_connect"
    case ("ipv6_connect", "ipv6", "2001:db8::2", "2001:db8::1"):
        fixtureCase = "ipv6_connect"
    case ("udp_send", "ipv4", "192.0.2.2", "192.0.2.1"):
        fixtureCase = "udp_send"
    case ("loopback_connect", "ipv4", "127.0.0.1", "127.0.0.1"):
        fixtureCase = "loopback_connect"
    case ("private_address_connect", "ipv4", "10.0.0.2", "10.0.0.1"):
        fixtureCase = "private_address_connect"
    case ("link_local_connect", "ipv4", "169.254.100.2", "169.254.100.1"):
        fixtureCase = "link_local_connect"
    case ("metadata_address_connect", "ipv4", "169.254.169.253", "169.254.169.254"):
        fixtureCase = "metadata_address_connect"
    case ("public_address_connect", "ipv4", "198.51.100.2", "198.51.100.1"):
        fixtureCase = "public_address_connect"
    case ("dns_plaintext", "ipv4", "192.0.2.2", "192.0.2.53"):
        fixtureCase = "dns_plaintext"
    default:
        throw LinuxVzNetworkEvidencePayloadError.invalidSchema
    }
    let expectedAction = fixtureCase == "dns_plaintext" ? "dns_query" :
        (fixtureCase == "udp_send" ? "udp_send" : "tcp_connect")
    let udpActivity = fixtureCase == "udp_send" || fixtureCase == "dns_plaintext"
    let expectedProtocol = udpActivity ? "udp" : "tcp"
    let expectedSocketState = udpActivity ? "unconnected_bound" :
        (fixtureCase == "loopback_connect" ? "established" : "syn_sent")
    let expectedTargetPort: UInt64 = fixtureCase == "dns_plaintext" ? 53 :
        (fixtureCase == "loopback_connect" ? 40_552 : 443)
    guard Set(value.keys) == expectedKeys,
          value["schema_version"] as? String == linuxVzNetworkEvidencePayloadSchemaV1,
          value["network_action"] as? String == expectedAction,
          value["network_protocol"] as? String == expectedProtocol,
          value["network_socket_state"] as? String == expectedSocketState,
          networkEvidenceDecimal(value["network_target_port"]) == expectedTargetPort,
          let sourcePort = networkEvidenceDecimal(value["network_source_port"]),
          sourcePort > 0, sourcePort <= UInt64(UInt16.max),
          networkEvidenceDecimal(value["event_sequence_start"]) == 1,
          networkEvidenceDecimal(value["event_sequence_end"]) == 4,
          networkEvidenceDecimal(value["event_count"]) == 4,
          networkEvidenceDecimal(value["heartbeat_count"]) == 2,
          networkEvidenceDecimal(value["dropped_event_count"]) == 0,
          networkEvidenceDecimal(value["package_uid"]) == 65534,
          networkEvidenceDecimal(value["package_gid"]) == 65534,
          networkEvidenceDecimal(value["reaped_process_count"]) == 1,
          value["sensor_healthy"] as? Bool == true,
          value["evidence_truncated"] as? Bool == false,
          value["descendant_teardown_complete"] as? Bool == true,
          let events = value["events"] as? [[String: Any]], events.count == 4 else {
        throw LinuxVzNetworkEvidencePayloadError.invalidSchema
    }
    let expectedKinds = ["fork", "exec", udpActivity ? "sendto" : "connect", "exit"]
    var childPID: UInt64?
    var cgroupID: UInt64?
    var previousTimestamp: UInt64 = 0
    for (index, event) in events.enumerated() {
        guard Set(event.keys) == Set([
            "actor_pid", "cgroup_id", "kind", "sequence", "subject_pid", "timestamp_ns"
        ]),
        event["kind"] as? String == expectedKinds[index],
        networkEvidenceDecimal(event["sequence"]) == UInt64(index + 1),
        let actor = networkEvidenceDecimal(event["actor_pid"]), actor > 0,
        let subject = networkEvidenceDecimal(event["subject_pid"]), subject > 0,
        let cgroup = networkEvidenceDecimal(event["cgroup_id"]), cgroup > 0,
        let timestamp = networkEvidenceDecimal(event["timestamp_ns"]),
        timestamp > previousTimestamp else {
            throw LinuxVzNetworkEvidencePayloadError.invalidEvent
        }
        if index == 0 {
            guard actor != subject else { throw LinuxVzNetworkEvidencePayloadError.invalidEvent }
            childPID = subject
            cgroupID = cgroup
        } else {
            guard actor == childPID, subject == childPID, cgroup == cgroupID else {
                throw LinuxVzNetworkEvidencePayloadError.invalidEvent
            }
        }
        previousTimestamp = timestamp
    }
    let claims = LinuxVzTelemetryGuestObservationClaims(
        evidencePayloadSHA256: sha256(data),
        evidenceByteLength: UInt64(data.count),
        eventSequenceStart: 1,
        eventSequenceEnd: 4,
        eventCount: 4,
        heartbeatCount: 2,
        droppedEventCount: 0,
        sensorHealthy: true,
        evidenceTruncated: false,
        descendantTeardownComplete: true,
        observedTerminal: "observation_complete"
    )
    return LinuxVzNetworkEvidencePayload(
        canonicalJSON: data,
        payloadSHA256: claims.evidencePayloadSHA256,
        evidenceByteLength: UInt64(data.count),
        fixtureCase: fixtureCase,
        packageUID: 65534,
        packageGID: 65534,
        sourcePort: UInt16(sourcePort),
        claims: claims
    )
}

private func networkEvidenceDecimal(_ value: Any?) -> UInt64? {
    guard let text = value as? String, !text.isEmpty,
          text == "0" || (text.first != "0" && text.allSatisfy(\.isNumber)) else {
        return nil
    }
    return UInt64(text)
}

File: whoathere/crates/whoathere-macos-vm/src/linux_vz_network_evidence_payload.rs
use crate::{
    LinuxVzTelemetryConformanceCaseV1, LinuxVzTelemetryConformanceObservedTerminalV1,
    LinuxVzTelemetryGuestObservationClaimsV1, MacosLinuxVzTelemetryEvidenceErrorV1,
};
use serde::{Deserialize, Serialize};
use std::fmt;
use whoathere_artifact::Sha256Digest;

pub const LINUX_VZ_NETWORK_EVIDENCE_PAYLOAD_SCHEMA_V1: &str =
    "whoathere.linux_vz_network_evidence_payload.v1";
pub const LINUX_VZ_NETWORK_EVIDENCE_SERIAL_PREFIX_V1: &[u8] = b"WHOATHERE_GUEST_NETWORK_EVIDENCE ";
pub const MAX_LINUX_VZ_NETWORK_EVIDENCE_PAYLOAD_BYTES_V1: usize = 64 * 1024;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LinuxVzNetworkEvidencePayloadErrorV1 {
    Missing,
    Duplicate,
    LimitExceeded,
    InvalidJson,
    NonCanonical,
    InvalidSchema,
    InvalidEvent,
}

impl fmt::Display for LinuxVzNetworkEvidencePayloadErrorV1 {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(match self {
            Self::Missing => "linux_vz_network_evidence_missing",
            Self::Duplicate => "linux_vz_network_evidence_duplicate",
            Self::LimitExceeded => "linux_vz_network_evidence_limit_exceeded",
            Self::InvalidJson => "linux_vz_network_evidence_json_invalid",
            Self::NonCanonical => "linux_vz_network_evidence_noncanonical",
            Self::InvalidSchema => "linux_vz_network_evidence_schema_invalid",
            Self::InvalidEvent => "linux_vz_network_evidence_event_invalid",
        })
    }
}

impl std::error::Error for LinuxVzNetworkEvidencePayloadErrorV1 {}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LinuxVzNetworkEvidencePayloadV1 {
    canonical_json: Vec<u8>,
    payload_sha256: Sha256Digest,
    fixture_case: LinuxVzTelemetryConformanceCaseV1,
    package_uid: u32,
    package_gid: u32,
    source_port: u16,
}

impl LinuxVzNetworkEvidencePayloadV1 {
    pub fn canonical_json_v1(&self) -> &[u8] {
        &self.canonical_json
    }

    pub fn payload_sha256(&self) -> &Sha256Digest {
        &self.payload_sha256
    }

    pub fn fixture_case(&self) -> LinuxVzTelemetryConformanceCaseV1 {
        self.fixture_case
    }

    pub fn source_port(&self) -> u16 {
        self.source_port
    }

    pub fn package_uid(&self) -> u32 {
        self.package_uid
    }

    pub fn package_gid(&self) -> u32 {
        self.package_gid
    }

    pub fn guest_observation_claims_v1(
        &self,
    ) -> Result<LinuxVzTelemetryGuestObservationClaimsV1, MacosLinuxVzTelemetryEvidenceErrorV1>
    {
        LinuxVzTelemetryGuestObservationClaimsV1::new(
            self.payload_sha256.clone(),
            self.canonical_json.len() as u64,
            1,
            4,
            4,
            2,
            0,
            true,
            false,
            true,
            LinuxVzTelemetryConformanceObservedTerminalV1::ObservationComplete,
        )
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
struct NetworkEvidenceWireV1 {
    descendant_teardown_complete: bool,
    dropped_event_count: String,
    event_count: String,
    event_sequence_end: String,
    event_sequence_start: String,
    events: Vec<NetworkEventWireV1>,
    evidence_truncated: bool,
    fixture_case: String,
    heartbeat_count: String,
    network_action: String,
    network_family: String,
    network_protocol: String,
    network_socket_state: String,
    network_source: String,
    network_source_port: String,
    network_target: String,
    network_target_port: String,
    package_gid: String,
    package_uid: String,
    reaped_process_count: String,
    schema_version: String,
    sensor_healthy: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
struct NetworkEventWireV1 {
    actor_pid: String,
    cgroup_id: String,
    kind: String,
    sequence: String,
    subject_pid: String,
    timestamp_ns: String,
}

pub fn decode_linux_vz_network_evidence_from_serial_v1(
    serial: &[u8],
) -> Result<LinuxVzNetworkEvidencePayloadV1, LinuxVzNetworkEvidencePayloadErrorV1> {
    let mut payload = None;
    for raw_line in serial.split(|byte| *byte == b'\n') {
        let line = raw_line.strip_suffix(b"\r").unwrap_or(raw_line);
        let Some(candidate) = line.strip_prefix(LINUX_VZ_NETWORK_EVIDENCE_SERIAL_PREFIX_V1) else {
            continue;
        };
        if payload.replace(candidate).is_some() {
            return Err(LinuxVzNetworkEvidencePayloadErrorV1::Duplicate);
        }
    }
    decode_linux_vz_network_evidence_payload_v1(
        payload.ok_or(LinuxVzNetworkEvidencePayloadErrorV1::Missing)?,
    )
}

pub fn decode_linux_vz_network_evidence_payload_v1(
    payload: &[u8],
) -> Result<LinuxVzNetworkEvidencePayloadV1, LinuxVzNetworkEvidencePayloadErrorV1> {
    if payload.is_empty() || payload.len() > MAX_LINUX_VZ_NETWORK_EVIDENCE_PAYLOAD_BYTES_V1 {
        return Err(LinuxVzNetworkEvidencePayloadErrorV1::LimitExceeded);
    }
    let mut deserializer = serde_json::Deserializer::from_slice(payload);
    let wire = NetworkEvidenceWireV1::deserialize(&mut deserializer)
        .map_err(|_| LinuxVzNetworkEvidencePayloadErrorV1::InvalidJson)?;
    deserializer
        .end()
        .map_err(|_| LinuxVzNetworkEvidencePayloadErrorV1::InvalidJson)?;
    let canonical = serde_json_canonicalizer::to_vec(&wire)
        .map_err(|_| LinuxVzNetworkEvidencePayloadErrorV1::InvalidJson)?;
    if canonical != payload {
        return Err(LinuxVzNetworkEvidencePayloadErrorV1::NonCanonical);
    }
    let source_port = decimal_u64_v1(&wire.network_source_port)?;
    let fixture_case = match wire.fixture_case.as_str() {
        "ipv4_connect"
            if wire.network_family == "ipv4"
                && wire.network_source == "192.0.2.2"
                && wire.network_target == "192.0.2.1" =>
        {
            LinuxVzTelemetryConformanceCaseV1::Ipv4Connect
        }
        "ipv6_connect"
            if wire.network_family == "ipv6"
                && wire.network_source == "2001:db8::2"
                && wire.network_target == "2001:db8::1" =>
        {
            LinuxVzTelemetryConformanceCaseV1::Ipv6Connect
        }
        "udp_send"
            if wire.network_family == "ipv4"
                && wire.network_source == "192.0.2.2"
                && wire.network_target == "192.0.2.1" =>
        {
            LinuxVzTelemetryConformanceCaseV1::UdpSend
        }
        "loopback_connect"
            if wire.network_family == "ipv4"
                && wire.network_source == "127.0.0.1"
                && wire.network_target == "127.0.0.1" =>
        {
            LinuxVzTelemetryConformanceCaseV1::LoopbackConnect
        }
        _ => return Err(LinuxVzNetworkEvidencePayloadErrorV1::InvalidSchema),
    };
    let (expected_action, expected_protocol, expected_socket_state, expected_event_kind) =
        if fixture_case == LinuxVzTelemetryConformanceCaseV1::UdpSend {
            ("udp_send", "udp", "unconnected_bound", "sendto")
        } else if fixture_case == LinuxVzTelemetryConformanceCaseV1::LoopbackConnect {
            ("tcp_connect", "tcp", "established", "connect")
        } else {
            ("tcp_connect", "tcp", "syn_sent", "connect")
        };
    let expected_target_port = if fixture_case == LinuxVzTelemetryConformanceCaseV1::LoopbackConnect
    {
        40_552
    } else {
        443
    };
    if wire.schema_version != LINUX_VZ_NETWORK_EVIDENCE_PAYLOAD_SCHEMA_V1
        || wire.network_action != expected_action
        || wire.network_protocol != expected_protocol
        || wire.network_socket_state != expected_socket_state
        || decimal_u64_v1(&wire.network_target_port)? != expected_target_port
        || source_port == 0
        || source_port > u16::MAX as u64
        || decimal_u64_v1(&wire.event_sequence_start)? != 1
        || decimal_u64_v1(&wire.event_sequence_end)? != 4
        || decimal_u64_v1(&wire.event_count)? != 4
        || decimal_u64_v1(&wire.heartbeat_count)? != 2
        || decimal_u64_v1(&wire.dropped_event_count)? != 0
        || decimal_u64_v1(&wire.package_uid)? != 65534
        || decimal_u64_v1(&wire.package_gid)? != 65534
        || decimal_u64_v1(&wire.reaped_process_count)? != 1
        || !wire.sensor_healthy
        || wire.evidence_truncated
        || !wire.descendant_teardown_complete
        || wire.events.len() != 4
    {
        return Err(LinuxVzNetworkEvidencePayloadErrorV1::InvalidSchema);
    }
    let expected_kinds = ["fork", "exec", expected_event_kind, "exit"];
    let mut child_pid = None;
    let mut cgroup_id = None;
    let mut prior_timestamp = 0;
    for (index, event) in wire.events.iter().enumerate() {
        let actor = decimal_u64_v1(&event.actor_pid)?;
        let subject = decimal_u64_v1(&event.subject_pid)?;
        let cgroup = decimal_u64_v1(&event.cgroup_id)?;
        let timestamp = decimal_u64_v1(&event.timestamp_ns)?;
        if event.kind != expected_kinds[index]
            || decimal_u64_v1(&event.sequence)? != index as u64 + 1
            || actor == 0
            || subject == 0
            || cgroup == 0
            || timestamp <= prior_timestamp
        {
            return Err(LinuxVzNetworkEvidencePayloadErrorV1::InvalidEvent);
        }
        if index == 0 {
            if actor == subject {
                return Err(LinuxVzNetworkEvidencePayloadErrorV1::InvalidEvent);
            }
            child_pid = Some(subject);
            cgroup_id = Some(cgroup);
        } else if actor != child_pid.unwrap()
            || subject != child_pid.unwrap()
            || cgroup != cgroup_id.unwrap()
        {
            return Err(LinuxVzNetworkEvidencePayloadErrorV1::InvalidEvent);
        }
        prior_timestamp = timestamp;
    }
    Ok(LinuxVzNetworkEvidencePayloadV1 {
        canonical_json: canonical,
        payload_sha256: Sha256Digest::from_bytes(payload),
        fixture_case,
        package_uid: 65534,
        package_gid: 65534,
        source_port: source_port as u16,
    })
}

fn decimal_u64_v1(value: &str) -> Result<u64, LinuxVzNetworkEvidencePayloadErrorV1> {
    if value.is_empty()
        || (value != "0" && value.starts_with('0'))
        || !value.bytes().all(|byte| byte.is_ascii_digit())
    {
        return Err(LinuxVzNetworkEvidencePayloadErrorV1::InvalidSchema);
    }
    value
        .parse()
        .map_err(|_| LinuxVzNetworkEvidencePayloadErrorV1::InvalidSchema)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn payload() -> Vec<u8> {
        serde_json_canonicalizer::to_vec(&serde_json::json!({
            "descendant_teardown_complete": true,
            "dropped_event_count": "0",
            "event_count": "4",
            "event_sequence_end": "4",
            "event_sequence_start": "1",
            "events": [
                {"actor_pid":"10","cgroup_id":"21","kind":"fork","sequence":"1","subject_pid":"11","timestamp_ns":"100"},
                {"actor_pid":"11","cgroup_id":"21","kind":"exec","sequence":"2","subject_pid":"11","timestamp_ns":"200"},
                {"actor_pid":"11","cgroup_id":"21","kind":"connect","sequence":"3","subject_pid":"11","timestamp_ns":"300"},
                {"actor_pid":"11","cgroup_id":"21","kind":"exit","sequence":"4","subject_pid":"11","timestamp_ns":"400"}
            ],
            "evidence_truncated": false,
            "fixture_case": "ipv4_connect",
            "heartbeat_count": "2",
            "network_action": "tcp_connect",
            "network_family": "ipv4",
            "network_protocol": "tcp",
            "network_socket_state": "syn_sent",
            "network_source": "192.0.2.2",
            "network_source_port": "49152",
            "network_target": "192.0.2.1",
            "network_target_port": "443",
            "package_gid": "65534",
            "package_uid": "65534",
            "reaped_process_count": "1",
            "schema_version": LINUX_VZ_NETWORK_EVIDENCE_PAYLOAD_SCHEMA_V1,
            "sensor_healthy": true
        })).unwrap()
    }

    #[test]
    fn ipv4_payload_binds_exact_sinkhole_tuple() {
        let evidence = decode_linux_vz_network_evidence_payload_v1(&payload()).unwrap();
        assert_eq!(
            evidence.fixture_case(),
            LinuxVzTelemetryConformanceCaseV1::Ipv4Connect
        );
        assert_eq!(evidence.source_port(), 49152);
        assert_eq!(
            evidence
                .guest_observation_claims_v1()
                .unwrap()
                .evidence_payload_sha256(),
            evidence.payload_sha256()
        );
    }

    #[test]
    fn ipv4_payload_rejects_target_rebinding() {
        let mut value: serde_json::Value = serde_json::from_slice(&payload()).unwrap();
        value["network_target"] = serde_json::json!("169.254.169.254");
        let forged = serde_json_canonicalizer::to_vec(&value).unwrap();
        assert_eq!(
            decode_linux_vz_network_evidence_payload_v1(&forged),
            Err(LinuxVzNetworkEvidencePayloadErrorV1::InvalidSchema)
        );
    }

    #[test]
    fn ipv6_payload_binds_exact_documentation_sinkhole() {
        let mut value: serde_json::Value = serde_json::from_slice(&payload()).unwrap();
        value["fixture_case"] = serde_json::json!("ipv6_connect");
        value["network_family"] = serde_json::json!("ipv6");
        value["network_source"] = serde_json::json!("2001:db8::2");
        value["network_target"] = serde_json::json!("2001:db8::1");
        let encoded = serde_json_canonicalizer::to_vec(&value).unwrap();
        let evidence = decode_linux_vz_network_evidence_payload_v1(&encoded).unwrap();
        assert_eq!(
            evidence.fixture_case(),
            LinuxVzTelemetryConformanceCaseV1::Ipv6Connect
        );
    }

    #[test]
    fn udp_payload_binds_sendto_and_exact_documentation_sinkhole() {
        let mut value: serde_json::Value = serde_json::from_slice(&payload()).unwrap();
        value["fixture_case"] = serde_json::json!("udp_send");
        value["network_action"] = serde_json::json!("udp_send");
        value["network_protocol"] = serde_json::json!("udp");
        value["network_socket_state"] = serde_json::json!("unconnected_bound");
        value["events"][2]["kind"] = serde_json::json!("sendto");
        let encoded = serde_json_canonicalizer::to_vec(&value).unwrap();
        let evidence = decode_linux_vz_network_evidence_payload_v1(&encoded).unwrap();
        assert_eq!(
            evidence.fixture_case(),
            LinuxVzTelemetryConformanceCaseV1::UdpSend
        );
    }

    #[test]
    fn loopback_payload_binds_established_guest_sinkhole() {
        let mut value: serde_json::Value = serde_json::from_slice(&payload()).unwrap();
        value["fixture_case"] = serde_json::json!("loopback_connect");
        value["network_socket_state"] = serde_json::json!("established");
        value["network_source"] = serde_json::json!("127.0.0.1");
        value["network_target"] = serde_json::json!("127.0.0.1");
        value["network_target_port"] = serde_json::json!("40552");
        let encoded = serde_json_canonicalizer::to_vec(&value).unwrap();
        let evidence = decode_linux_vz_network_evidence_payload_v1(&encoded).unwrap();
        assert_eq!(
            evidence.fixture_case(),
            LinuxVzTelemetryConformanceCaseV1::LoopbackConnect
        );
    }
}
